File: sluice/monitor/guard.py
"""Provider-neutral guard for an existing agent loop.

Provider adapters convert each request's messages into neutral ``Message`` objects and call
``ingest``; they convert the model's reply and call ``review``, which checks every tool call
and returns the reply with blocked calls removed. The caller's loop then only ever sees
tool calls that passed policy.
"""

from __future__ import annotations

import json
from collections.abc import Sequence
from dataclasses import dataclass, field, replace
from typing import Any

from sluice.llm import Message
from sluice.monitor.middleware import CheckedCall, Monitor
from sluice.policy.decision import Decision


class SluiceBlocked(RuntimeError):
    """Raised by a guard configured with ``raise_on_block=True``."""

    def __init__(self, decisions: Sequence[Decision]) -> None:
        self.decisions = list(decisions)
        super().__init__("\n\n".join(d.explain() for d in self.decisions))


@dataclass(frozen=True)
class Review:
    message: Message  # the reply with blocked tool calls removed
    allowed: list[CheckedCall] = field(default_factory=list)
    blocked: list[CheckedCall] = field(default_factory=list)

    @property
    def notice(self) -> str:
        """Text appended to the reply so the model and the user see what happened."""
        if not self.blocked:
            return ""
        lines = [f"[sluice blocked {c.call.name}]\n{c.decision.explain()}" for c in self.blocked]
        return "\n\n".join(lines)


def parse_output(content: str) -> Any:
    """Tool results arrive as text; JSON objects/arrays are labelled per field."""
    s = content.strip()
    if s[:1] in "[{":
        try:
            return json.loads(s)
        except ValueError:
            pass
    return content


class Guard:
    def __init__(self, monitor: Monitor, *, raise_on_block: bool = False) -> None:
        self.monitor = monitor
        self.raise_on_block = raise_on_block
        self._seen: set[tuple[str, str]] = set()
        self._pending: dict[str, CheckedCall] = {}

    def ingest(self, messages: Sequence[Message]) -> None:
        """Label everything new in the conversation. Safe to call with the full history on
        every request: each message is recorded once."""
        for m in messages:
            if m.role in ("system", "user"):
                key: tuple[str, str] = (m.role, m.content)
                if m.content and key not in self._seen:
                    self._seen.add(key)
                    self.monitor.observe(m.content, m.role)
            elif m.role == "tool":
                key = ("tool", m.tool_call_id or f"anon:{m.name}:{m.content}")
                if key in self._seen:
                    continue
                self._seen.add(key)
                checked = self._pending.pop(m.tool_call_id or "", None)
                output = parse_output(m.content)
                if checked is not None:
                    self.monitor.record_output(checked, output)
                else:
                    self.monitor.record_unreviewed_output(m.name, output)
            # Assistant text is the model's own output; its tool calls are reviewed in review().

    def review(self, reply: Message) -> Review:
        allowed: list[CheckedCall] = []
        blocked: list[CheckedCall] = []
        for tc in reply.tool_calls:
            checked = self.monitor.check(tc)
            if checked.decision.allowed:
                allowed.append(checked)
                self._pending[tc.id] = checked
            else:
                blocked.append(checked)
        if blocked and self.raise_on_block:
            raise SluiceBlocked([c.decision for c in blocked])
        kept = tuple(c.call for c in allowed)
        review = Review(replace(reply, tool_calls=kept), allowed, blocked)
        if blocked:
            content = "\n\n".join(x for x in (reply.content, review.notice) if x)
            review = replace(review, message=replace(review.message, content=content))
        return review

File: sdk/typescript/src/guard.ts
/**
 * A drop-in guard for a TypeScript tool-calling loop.
 *
 * `ingest` labels new conversation content; `review` checks the model's tool calls and returns
 * only the ones that pass policy, with an explanation for the blocked ones. The caller's loop
 * then only ever executes calls sluice allowed.
 */

import { Label } from "./labels.js";
import { Monitor, type CheckedCall, type ToolCall } from "./monitor.js";
import type { Decision } from "./policy.js";

export type Role = "system" | "user" | "assistant" | "tool";

export interface Message {
  role: Role;
  content?: string;
  toolCalls?: ToolCall[];
  toolCallId?: string;
  name?: string;
}

export class SluiceBlocked extends Error {
  constructor(readonly decisions: Decision[]) {
    super(decisions.map((d) => d.explain()).join("\n\n"));
    this.name = "SluiceBlocked";
  }
}

export interface Review {
  readonly allowed: CheckedCall[];
  readonly blocked: CheckedCall[];
  /** Text describing blocked calls, to append to the reply the model and user see. */
  readonly notice: string;
}

export function parseOutput(content: string): unknown {
  const s = content.trim();
  if (s.startsWith("[") || s.startsWith("{")) {
    try {
      return JSON.parse(s);
    } catch {
      /* fall through */
    }
  }
  return content;
}

export class Guard {
  private seen = new Set<string>();
  private pending = new Map<string, CheckedCall>();

  constructor(
    readonly monitor: Monitor,
    readonly options: { raiseOnBlock?: boolean } = {},
  ) {}

  /** Label everything new in the conversation. Safe to call with the full history each turn. */
  ingest(messages: Message[]): void {
    for (const m of messages) {
      if (m.role === "system" || m.role === "user") {
        const key = `${m.role}:${m.content ?? ""}`;
        if (m.content && !this.seen.has(key)) {
          this.seen.add(key);
          this.monitor.observe(m.content, m.role);
        }
      } else if (m.role === "tool") {
        const key = `tool:${m.toolCallId ?? `${m.name}:${m.content}`}`;
        if (this.seen.has(key)) continue;
        this.seen.add(key);
        const checked = m.toolCallId ? this.pending.get(m.toolCallId) : undefined;
        const output = parseOutput(m.content ?? "");
        if (checked) {
          this.monitor.recordOutput(checked, output);
          this.pending.delete(m.toolCallId!);
        } else {
          this.monitor.recordUnreviewedOutput(m.name ?? "", output);
        }
      }
    }
  }

  review(reply: Message): Review {
    const allowed: CheckedCall[] = [];
    const blocked: CheckedCall[] = [];
    for (const tc of reply.toolCalls ?? []) {
      const checked = this.monitor.check(tc);
      if (checked.decision.allowed) {
        allowed.push(checked);
        this.pending.set(tc.id, checked);
      } else {
        blocked.push(checked);
      }
    }
    if (blocked.length && this.options.raiseOnBlock) {
      throw new SluiceBlocked(blocked.map((c) => c.decision));
    }
    const notice = blocked.map((c) => `[sluice blocked ${c.call.name}]\n${c.decision.explain()}`).join("\n\n");
    return { allowed, blocked, notice };
  }
}

export { Label };
